File: 백준/Gold/2580. 스도쿠/스도쿠.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "ex.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const graph = [];
const blank = [];

for (let i = 0; i < 9; i++) {
  graph.push(input[i].split(" ").map(Number));
}

for (let i = 0; i < 9; i++) {
  for (let j = 0; j < 9; j++) {
    if (graph[i][j] === 0) {
      blank.push([i, j]);
    }
  }
}

const row = (x, a) => {
  for (let i = 0; i < 9; i++) {
    if (a === graph[x][i]) {
      return false;
    }
  }
  return true;
};

const column = (y, a) => {
  for (let i = 0; i < 9; i++) {
    if (a === graph[i][y]) {
      return false;
    }
  }
  return true;
};

const square = (x, y, a) => {
  const nx = Math.floor(x / 3) * 3;
  const ny = Math.floor(y / 3) * 3;
  for (let i = nx; i < nx + 3; i++) {
    for (let j = ny; j < ny + 3; j++) {
      if (a === graph[i][j]) {
        return false;
      }
    }
  }
  return true;
};

const dfs = (depth) => {
  if (depth === blank.length) {
    for (const g of graph) {
      console.log(g.join(" "));
    }
    process.exit();
  }

  for (let i = 1; i < 10; i++) {
    const [x, y] = blank[depth];

    if (row(x, i) && column(y, i) && square(x, y, i)) {
      graph[x][y] = i;
      dfs(depth + 1);
      graph[x][y] = 0;
    }
  }
};

dfs(0);

File: 백준/Gold/2580. 스도쿠/스도쿠.py
graph = [list(map(int, input().split())) for _ in range(9)]

def row(x, a):
    for i in range(9):
        if a == graph[x][i]:
            return False
    return True

def col(y, a):
    for i in range(9):
        if a == graph[i][y]:
            return False
    return True

def square(x, y, a):
    nx = x // 3 * 3
    ny = y // 3 * 3
    for i in range(3):
        for j in range(3):
            if a == graph[nx+i][ny+j]:
                return False
    return True

def dfs(depth):
    if depth == len(blank):
        for idx in range(9):
            print(*graph[idx])
        exit(0)
    for i in range(1, 10):
        x = blank[depth][0]
        y = blank[depth][1]

        if row(x, i) and col(y, i) and square(x, y, i):
            graph[x][y] = i
            dfs(depth + 1)
            graph[x][y] = 0
blank = []
for i in range(9):
    for j in range(9):
        if graph[i][j] == 0:
            blank.append((i, j))

dfs(0)